### PR TITLE
create_run_args copy-by-ref fix

### PR DESCRIPTION
**Pull request by ankona**
Add a deep copy to the batch & run settings `__init__` methods. This ensures that changes to collections after pass-by-ref are not unexpectedly modified after creating run/batch settings instances.

### CODE REVIEW

File: smartsim/settings/base.py [hunk 1]
```diff
@@ -78,19 +77,39 @@ def __init__(
...
-        if container:
-            self.exe = [exe]
-        else:
-            self.exe = [expand_exe_path(exe)]
-
-        self.exe_args = self._set_exe_args(exe_args)
-        self.run_args = init_default({}, run_args, dict)
-        self.env_vars = init_default({}, env_vars, dict)
+        self.exe = [exe] if container else [expand_exe_path(exe)]
+        self.exe_args = exe_args
```
[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/settings/base.py:81]: Similar "expected `T` but got `T | None`"  for `run_args` and `env_vars`

File: smartsim/settings/base.py [hunk 2]
```diff
@@ -78,19 +77,39 @@ def __init__(
...
+    @exe_args.setter
+    def exe_args(self, value: t.Union[str, t.List[str], None]) -> None:
+        self._exe_args = self._build_exe_args(value)
+
+    @property
+    def run_args(self) -> t.Dict[str, t.Optional[str]]:
+        return self._run_args
+    
+    @run_args.setter
+    def run_args(self, value: t.List[str]) -> None:
```
[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/settings/base.py:102]: same guess for the `env_vars` setter

**Review by MattToast — APPROVED** [review]
LGTM!! A couple of small typing, but other wise looks great!

Approving in advance!!

File: smartsim/settings/base.py [hunk 3]
```diff
@@ -477,11 +496,14 @@ def set(self, arg: str, value: t.Optional[str] = None, condition: bool = True) -
             logger.warning(f"Overwritting argument '{arg}' with value '{value}'")
         self.run_args[arg] = value
 
-    def _set_exe_args(self, exe_args: t.Optional[t.Union[str, t.List[str]]]) -> t.List[str]:
+    @staticmethod
+    def _build_exe_args(exe_args: t.Optional[t.Union[str, t.List[str]]]) -> t.List[str]:
+        """ Convert exe_args input to a desired collection format"""
```
[the remark this answers is not in the document]
Comment (ankona) [reply smartsim/settings/base.py]: removed extraneous spacing

File: smartsim/settings/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ankona) [reply smartsim/settings/base.py:102]: copypasta monster fail. tyvm. fixed.

File: smartsim/settings/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ankona) [reply smartsim/settings/base.py]: Fixed

**Review by mellis13 — APPROVED** [review]
One small question/comment -- approving in advance.  Thanks!

File: smartsim/settings/base.py [hunk 4]
```diff
@@ -36,12 +37,10 @@ class RunSettings:
     def __init__(
         self,
         exe: str,
-        exe_args: t.Optional[t.Union[str, t.List[str]]] = None,
+        exe_args: t.Optional[t.List[str]] = None,
```
[the remark this answers is not in the document]
Comment (ankona) [reply smartsim/settings/base.py]: I've undone the removal of the str as an option for setting exe_args.

**Comment by ankona** [comment]
@mellis13 

I've rebased off the latest develop branch (internal typehints were added) and did some tweaks to come inline with expectations. Mainly, the dictionaries for settings are changed to `t.Dict[str, *t.Optional*[str]]` (optional values).

I also updated some tests and functions to handle the automatic stringification of configuration values when stored (e.g. internally, all settings are t.Dict[str, str] even if you want something to be an integer, like cpus_per_node). Many tests existed that pass a string in and expected a ValueError to be thrown. You'll see some unsightly `str(int(value))` to ensure those same errors occurred.

File: smartsim/settings/slurmSettings.py [hunk 5]
```diff
@@ -201,7 +201,7 @@ def set_memory_per_node(self, memory_per_node: int) -> None:
         :param memory_per_node: Amount of memory per node in megabytes
         :type memory_per_node: int
         """
-        self.run_args["mem"] = f"{int(memory_per_node)}M"
+        self.run_args["mem"] = f"{str(int(memory_per_node))}M"
```
Comment (MattToast) [inline smartsim/settings/slurmSettings.py:204]: Is the str cast needed here to satisfy the typechecker? Should be redundant as it is cast implicitly by the f-str

**Review by MattToast — APPROVED** [review]
LGTM! One last question, but nothing worth holding approval over

**Review by mellis13 — APPROVED** [review]

File: smartsim/settings/slurmSettings.py (hunk 5 above)
Comment (ankona) [reply smartsim/settings/slurmSettings.py:204]: doh!

Comment (ankona) [reply smartsim/settings/slurmSettings.py]: reverted these changes

**Pull request merged**